doc/design/wallet/wallet-design.md

```diff
@@ -0,0 +1,724 @@
+
+                                           /// Create the wallet storage (For example, database creation)
+                                           ///
+                                           /// #Params
+                                           /// name: wallet storage name (the same as wallet name)
+                                           /// config: wallet storage config (For example, database config)
+                                           /// credentials: wallet storage credentials (For example, database credentials)
+                                           create: Option<extern fn(name: *const c_char,
+                                                                    config: *const c_char,
+                                                                    credentials: *const c_char) -> ErrorCode>,
+
+                                           /// Delete the wallet storage (For example, database deletion)
+                                           ///
+                                           /// #Params
+                                           /// name: wallet storage name (the same as wallet name)
+                                           /// config: wallet storage config (For example, database config)
+                                           /// credentials: wallet storage credentials (For example, database credentials)
+                                           delete: Option<extern fn(name: *const c_char,
+                                                                    config: *const c_char,
+                                                                    credentials: *const c_char) -> ErrorCode>,
+
+                                           /// Open the wallet storage (For example, opening database connection)
+                                           ///
+                                           /// #Params
+                                           /// name: wallet storage name (the same as wallet name)
+                                           /// config: wallet storage config (For example, database config)
+                                           /// runtime_config: wallet storage runtime config (For example, connection config)
+                                           /// credentials: wallet storage credentials (For example, database credentials)
+                                           /// storage_handle_p: pointer to store opened storage handle
+                                           open: Option<extern fn(name: *const c_char,
+                                                                  config: *const c_char,
+                                                                  runtime_config: *const c_char,
+                                                                  credentials: *const c_char,
+                                                                  storage_handle_p: *mut i32) -> ErrorCode>,
```

Comment:
May we make this storage handle *mut c_void. I would like to give storage implementer ability to put there pointer to whatever object it is the best for his implementation (for eg. pointer to the storage object or struct holding the internal state).